src/lib/libast/path/pathbin.c

```diff
@@ -19,28 +19,30 @@
  *                     Phong Vo <phongvo@gmail.com>                     *
  *                                                                      *
  ***********************************************************************/
-/*
- * Glenn Fowler
- * AT&T Bell Laboratories
- *
- * return current PATH
- */
+//
+//  Glenn Fowler
+//  AT&T Bell Laboratories
+//
 #include "config_ast.h"  // IWYU pragma: keep
 
 #include <string.h>
 
 #include "ast.h"
 
+//  Return current PATH
+//
 char *pathbin(void) {
-    char *bin;
+    char *path;
+    // Used to restore value of saved astconf path from earlier function call
+    static char *astconf_path;
 
-    static char *val;
-
-    if ((!(bin = getenv("PATH")) || !*bin) && !(bin = val)) {
-        if (!*(bin = astconf("PATH", NULL, NULL)) || !(bin = strdup(bin))) {
-            bin = "/bin:/usr/bin:/usr/local/bin";
+    path = getenv("PATH");
+    if ((!path || !*path) && !(path = astconf_path)) {
```

Comment:
And here we have an example why such assignments are dangerous. Note that `astconf_path` is uninitialized at this juncture which means the condition will be true or false depending on whatever random garbage is on the stack when this function is called. It looks like you intended to drop this assignment and test from the `if()` since you moved it into the body.